### ISSUE TITLE
Well-Typedness of Function Expressions

### ISSUE THREAD

**Issue by jg-rp** [jsonpath-standard/jsonpath-compliance-test-suite#54]
Using function extensions currently defined by the standard, it's not possible to test some valid JSONPath queries. Specifically, no standard function extension returns a _NodesType_ result, and no standard function extension accepts a _LogicalType_ parameter.

For example, if `func()` were a function extension defined to accept a _LogicalType_ parameter and return a _LogicalType_ result, the following queries would be valid.

- `$[func(@.a == @.b)]` 
- `$[func(@.*)]`  ([section 2.4.2 - Type Conversion](https://datatracker.ietf.org/doc/html/draft-ietf-jsonpath-base-21#name-type-conversion))

The table in [section 2.4.9] contains more combinations of parameter/result types that can't currently be tested.

The question is, should this test suite define non-standard function extensions in order to cover these cases?

If so, how could that be done? And should we attempt to maintain backwards compatibility for dependants that don't implement non-standard function extensions (that is, keep such test cases out of `cts.json`)?

See also the conversation on PR #51, which attempted to add a query similar to the first example above.

**Comment by glyn** [comment]
> The question is, should this test suite define non-standard function extensions in order to cover these cases?

I don't think it should because presumably that would mean implementations would need to implement those non-standard extensions in order to pass the test suite.

I suppose one way round this would be to register standard function extensions for the purpose of testing.

**Comment by f3ath** [comment]
FWIW my implementation supports custom functions and provides a few non-standard ones as examples and for testing purposes, including
- `siblings()` which is `(NodesType) -> NodesType`
- `xor()` which is `(LogicalType, LogicalType) -> LogicalType`
